README.md

# miRBaseVersions.db

The _miRBaseVersions.db_ R package has been developed to provide an easy accessible repository for several different miRBase release versions ([miRBase](http://www.mirbase.org)).
It contains 22 different miRBase release versions and implements the `select` interface.

* [Introduction](#introduction)
* [Use Cases](#use-cases)
   - [Function `keytypes`](#function-keytypes)
   - [Function `columns`](#function-columns)
   - [Function `keys`](#function-keys)
   - [Function `select`](#function-select)
* [Additional Information](#additional-information)

## Introduction

The _miRBaseVersions.db_ package is an annotation package which includes 
mature miRNA names from 22 miRBase release versions. Due to ongoing growth and 
changes with each release miRNA names can have different names in different 
versions or even are not listed as valid miRNAs anymore. This annotation package
serves as a repository and can be used for quick lookup for mature miRNA names. 
The _miRBaseVersions.db_ package has implemented the AnnotationDbi-select 
interface. By implementing this `select` interface the user is able to use 
the same methods as for any other annotation package.

The main four implemented methods are 

- `columns`, presents the values one can retrieve in the final result,
- `keytypes`, which presents the tables that can be used in this package,
- `keys`, is used to get viable keys of a particular `keytype` and
- `select`, which is used to extract data from the annotation package by using 
values provided by the other three methods.

To load the package and gain access to the functions just run the 
following command:

```r
library(miRBaseVersions.db)
```

### Database information

The data is the _miRNAmeConverter_ package is stored in an SQLite database. All entries contained in the database were downloaded from the [miRBase ftp-site](ftp://mirbase.org/pub/mirbase/). The following versions are available:


|miRBase Version |Release Date	  | # Mature entries|
|---------------:|---------------:|----------------:|
|	6.0	         |	04/05         |	1650	        |
|	7.1	         |	10/05         |	3424	        |
|	8.0	         |	02/06         |	3518	        |
|	8.1	         |	05/06         |	3963	        |
|	8.2	         |	07/06         |	4039	        |
|	9.0	         |	10/06         |	4361	        |
|	9.1	         |	02/07         |	4449	        |
|	9.2	         |	05/07         |	4584	        |
|	10.0         |	08/07         |	5071	        |
|	10.1         |	12/07         |	5395	        |
|	11.0         |	04/08         |	6396	        |
|	12.0         |	09/08         |	8619	        |
|	13.0         |	03/09         |	9539	        |
|	14.0         |	09/09         |	10883	        |
|	15.0         |	04/10         |	14197	        |
|	16.0         |	08/10         |	15172	        |
|	17.0         |	04/11         |	16772	        |
|	18.0         |	11/11         |	18226	        |
|	19.0         |	08/12         |	21264	        |
|	20.0         |	06/13         |	24521	        |
|	21.0         |	06/14         |	28645	        |
|	22.0         |	03/18         |	38589	        |


from 271 organisms.

## Use Cases

### Function `keytypes`

Use this function to receive table names from where data can be retrieved:
```r
keytypes(miRBaseVersions.db);
```
The output lists `r length(keytypes(miRBaseVersions.db))` tables where each one
of them can be queried. The keytype "MIMAT" is the main table containing all
records from all supported miRBase release versions. Keytypes starting with 
the prefix "VW-MIMAT" are so called SQL views. For example the keytype
"VW-MIMAT-22.0" is an SQL view from the "MIMAT" table which only holds records
from miRBase version 22.0.

### Function `columns`

Use the `columns` function to retreive information about the kind of variables
you can retrieve in the final output:
```r
columns(miRBaseVersions.db);
```
All `r length(columns(miRBaseVersions.db))` columns are available for all 
`r length(keytypes(miRBaseVersions.db))` keytypes.

### Function `keys`

The `keys` function returns all viable keys of a praticular `keytype`. The 
following example retrieves all possible keys for miRBase release version 6.0.
```r
k = head(keys(miRBaseVersions.db, keytype = "VW-MIMAT-6.0"));
k
```

### Function `select`

The `select` function is used to extract data. As input values the function
takes outputs received from the other three functions `keys`, 
`columns` and `keytypes`.  
For exmaple to extract all information about the mature 
accession 'MIMAT0000092' we can run the following command:
```r
result = select(miRBaseVersions.db, 
                keys = "MIMAT0000092", 
                keytype = "MIMAT", 
                columns = "*")
result
```
As we can see the result returns all miRNA names the accession had among the different miRBase releases. If we for example only want to extract the fields for 'accession', 'name' and 'version' we simply run the following command:
```r
result = select(miRBaseVersions.db, 
                keys = "MIMAT0000092", 
                keytype = "MIMAT", 
                columns = c("ACCESSION", "NAME", "VERSION"))
result
```
In comparison to the previous output with parameter `columns = "*"` this time
only the selected columns were returned.

## Additional information

### Packages loaded via namespace
The following packages are used in the `miRBaseVersions.db` package: 

* AnnotationDbi_1.32.3
* DBI_0.3.1 [visit on CARN][http://CRAN.R-project.org/package=RSQLite]
* RSQLite_1.0.0 [visit on CARN](http://CRAN.R-project.org/package=DBI)
* gtools_3.5.0 [visit on CRAN](https://CRAN.R-project.org/package=gtools)

### Future Aspects
This database can only be of good use if it will be kept up to date.
Therefore we plan to include new miRBase releases as soon as possible.


inst/NEWS
**************************************************
*
* Initial Release
*
**************************************************

NEW FEATURES

    o Added miRBase release version 22, which was released on March 12th, 2018

o Future plans
   - Include miRBase release versions from 1.0 to 5.1 and 7.0


R/zzz.R
datacache = new.env(hash=TRUE, parent=emptyenv())

# mirbaseversions_dbconn = function() dbconn(datacache)
# mirbaseversions_dbfile = function() dbfile(datacache)

mirbaseversionsORGANISM = "Multiple"

#' @docType package
#' @name miRBaseVersions.db
#' @title miRBaseVersions.db: miRNA name collection of 22 different miRBase
#' release versions.
#' @aliases columns keys keytypes select
#' @details This annotation package comprises mature miRNA names from
#' 22 different miRBase versions. It contains one main table holding all miRNAs
#' and one view for each version, such as 'vw-mimat-22.0' for mature miRNA names
#' from version 22.0.
#' @import AnnotationDbi methods DBI RSQLite gtools
#' @author Stefan Haunsberger \email{stefanhaunsberger@rcsi.ie}
#'
#' @description object of \code{MiRBaseVersionsDb} class holds the sqlite
#' database connection, and extends \code{AnnotationDb} class from
#' AnnotationDbi package. \code{columns}, \code{keys}, \code{keytypes} and
#' \code{select} methods allow access to database tables and retrieval of
#' miRNA target information.
#'
#' \code{select} is for querying the database to retrieve information about
#' mature miRNA names from selected versions.
#' @usage columns(x)
#' keytypes(x)
#' keys(x, keytype)
#' select(x, keys, columns, keytype)
#' @param x the \code{miRBaseVersions.db}
#' @param keytype represents the table from which data shall be received.
#' All possible keytypes can be viewed by using the \code{keytypes} method.
#' @param keys the accession name of mature miRNAs. All possible keys
#' (miRNAs) are returned by using the \code{keys} method.
#' @param columns that can be returned for each miRNA. All possible columns
#' can be shown by using the \code{columns} method.
#' @param libname Name of library (automatically passed on at start up)
#' @param pkgname Package name (automatically passed on at start up)
#' @return string vectors, for \code{select} a data.frame with selected
#' columns.
#' @examples
# #first load the annotations
# require(miRBaseVersions.db)
#' # see all available tables
#' keytypes(miRBaseVersions.db)
#' # see column names
#' columns(miRBaseVersions.db)
NULL

# The .onLoad will be called when the package is loaded.
.onLoad = function(libname, pkgname) {

    # library(methods)
    # Connect to the SQLite database
    sPkgname = sub(".db$","",pkgname);
    ## Database file
    dbfile = system.file("extdata", paste0(sPkgname, ".sqlite"),
                         package = pkgname, lib.loc=libname)
    assign("dbfile", dbfile, envir = datacache);
    ## Establish database connection
    dbconn = AnnotationDbi::dbFileConnect(dbfile);
    assign("dbconn", dbconn, envir=datacache);

    ########################################################################
    # Define class

#     if(!exists("MiRBaseVersions", mode="class")) {
#         # stop("first statement ... !!!>!>!>!>!>")
#         if (!file.exists("R/ClassDef.R")) {
#             stop("second statement ... !!>!<!>!><!")
#             # source("R/ClassDef.R")
#         }
#         sys.source(file = "R/ClassDef.R", envir = environment())
#     }

    # @title Database class
    # @aliases MiRBaseVersionsDb columns keys keytypes select
    # @description object of \code{MiRBaseVersionsDb} class holds the sqlite
    # database connection, and extends \code{AnnotationDb} class from
    # AnnotationDbi package. \code{columns}, \code{keys}, \code{keytypes} and
    # \code{select} methods allow access to database tables and retrieval of
    # miRNA target information.
    #
    # \code{select} is for querying the database to retrieve information about
    # mature miRNA names from selected versions.
    # @usage columns(x)
    # keytypes(x)
    # keys(x, keytype)
    # select(x, keys, columns, keytype)
    # @param x the \code{miRBaseVersions.db}
    # @param keytype represents the table from which data shall be received.
    # All possible keytypes can be viewed by using the \code{keytypes} method.
    # @param keys the accession name of mature miRNAs. All possible keys
    # (miRNAs) are returned by using the \code{keys} method.
    # @param columns that can be returned for each miRNA. All possible columns
    # can be shown by using the \code{columns} method.
    # @return string vectors, for \code{select} a data.frame with selected
    # columns.
    # @author Stefan Haunsberger \email{stefanhaunsberger@rcsi.ie}
    # @examples
    # #first load the annotations
    # require(miRBaseVersions.db)
    # #see all available tables
    # keytypes(miRBaseVersions.db)
    # @export MiRBaseVersionsDb
    .MiRBaseVersionsDb = setRefClass(
        Class = "MiRBaseVersionsDb",
        contains = "AnnotationDb"
    )

    # @title Lower case columns function implementation
    # @details This function reads column names from the view of the
    # current version in lower case. All visible keytypes have the same columns.
    # @param MiRBaseVersionsDb object reference
    # @return Character vector with lower case column names.
    .getLCColnames = function(x) {
        # Retrieve tables names from database
        con = AnnotationDbi::dbconn(x);
        ## Only select columsn from one view
        # cols = grep("vw-mimat-[0-9]+\\.[0-9]$|organism",
        #               dbListTables(con), value = TRUE);
        # cols = grep("vw-mimat-[0-9]+\\.[0-9]$", dbListTables(con),
        #               value = TRUE);
        cols = (DBI::dbGetQuery(con,
                                "PRAGMA table_info('vw-mimat-22.0')"))$name;
        return(cols);
    }

    # @title Internal \code{columns} function implementation
    # @details This function converts lower case column names (received from
    # \code{\link{.getLCColNames}}) into upper case.
    # @param MiRBaseVersionsDb object reference
    # @return Character vector containing upper case column names
    .cols = function(x)
    {
        cols = .getLCColnames(x);
        cols = toupper(cols);
        return(cols);
    }

    # @title Internal \code{keytypes} implementation
    # @details The \code{.getTableNames} function reads table names from the
    # database. Two types of tables will be returned, the general mimat table
    # and views respectively for each miRBase version.
    # @param MiRBaseVersionsDb object reference
    # @return Character vector where the values are lower name and the 'name'
    # attribute are upper name table names.
    .getTableNames = function(x)
    {
        con = AnnotationDbi::dbconn(x);
        ## Receive table names
        tables = grep("vw-mimat-[0-9]+\\.[0-9](-hsa)?$|^\\bmimat\\b",
                      DBI::dbListTables(con), value = TRUE);
        names(tables) = toupper(tables);
        return(tables);
    }

    # @title Internal \code{keys} function implementation
    # @details The \code{.keys} function implementation reads the accessions
    # of the desired keytype.
    # @param MiRBaseVersionsDb object reference
    # @return Character vector with accession names from specified
    # \code{keytype}
    .keys = function(x, keytype)
    {
        ## translate keytype back to table name
        tabNames = .getTableNames(x);
        # lckeytype = names(tabNames[tabNames %in% keytype]);
        lckeytype = tabNames[names(tabNames) %in% keytype];
        if (length(lckeytype) == 0) {
            # message(sprintf("keytype '%s' not present.", keytype));
            stop(paste("keytype", keytype, "does not exist.",
                    "Please use method 'keytypes()' to check out the keytypes."
            ));
        }
        ## get a connection
        con = AnnotationDbi::dbconn(x);
        sql = character();
        #     if (length(grep("^vw.+", lckeytype, ignore.case = TRUE)) > 0) {
        #         sql = paste0("SELECT accession FROM \"", lckeytype, "\"");
        #     } else if (lckeytype == "organism") {
        #         sql = paste0("SELECT organism FROM \"", lckeytype, "\"");
        #     }
        sql = paste0("SELECT accession FROM \"", lckeytype, "\"");
        ## Execute SQL statement
        res = DBI::dbGetQuery(con, sql);
        res = as.vector(t(res));
        return(res);
    }


    .select = function(x, keys, columns, keytype) {
        ## translate keytype back to table name
        tabNames = .getTableNames(x);
        keytypeLC = tabNames[names(tabNames) %in% keytype];
        if (length(keytypeLC) == 0) {
            stop("Keytype not valid. Please check out 'keytypes'.");
        }
        ## Get columns
        colsSQL = character();
        colsLC = .getLCColnames(x);
        if (any(columns != "*")) {
            colsUC = .cols(x);
            cols = colsLC[colsUC %in% columns];
            colsSQL = paste(cols, collapse = ", ");
        } else {
            colsSQL = "*";
        }
        ## get the connection
        con = AnnotationDbi::dbconn(x);
        sql = sprintf(paste("SELECT %s FROM `%s`",
                            "WHERE UPPER(accession) IN ('%s')"),
                colsSQL, keytypeLC,
                paste(toupper(keys), collapse = "', '"));
        ## Execute SQL statement
        res = DBI::dbGetQuery(con, sql);
        colnames(res) = toupper(colnames(res));
        return(res)
    }

    # @rdname MiRBaseVersionsDb-class
    #' @rdname miRBaseVersions.db
    #' @exportMethod columns
    setMethod(
        f = "columns",
        signature = "MiRBaseVersionsDb",
        definition = function(x) {
            return(.cols(x));
        }
    )

    # @rdname MiRBaseVersionsDb-class
    #' @rdname miRBaseVersions.db
    #' @exportMethod keytypes
    setMethod(
        f = "keytypes",
        signature = "MiRBaseVersionsDb",
        definition = function(x) {
            tables = names(.getTableNames(x))
            ## Order values
            tables = tables[gtools::mixedorder(tables, decreasing = TRUE)];
            return(tables);
        }
    )

    # @rdname MiRBaseVersionsDb-class
    #' @rdname miRBaseVersions.db
    #' @exportMethod keys
    setMethod(
        f = "keys",
        signature = "MiRBaseVersionsDb",
        definition = function(x, keytype) {
            return(.keys(x, keytype));
        }
    )

    # @rdname MiRBaseVersionsDb-class
    #' @rdname miRBaseVersions.db
    #' @exportMethod select
    setMethod(
        f = "select",
        signature = "MiRBaseVersionsDb",
        definition = function(x, keys, columns = "*", keytype = "") {
            .select(x, keys = keys, columns = columns, keytype = keytype);
        }
    )



    ########################################################################

    # Create the OrgDb object
    ## Use information stored in the metadata table
    txdb = AnnotationDbi::loadDb(dbfile,
                                 packageName = pkgname);
    dbNewname = AnnotationDbi:::dbObjectName(pkgname, "MiRBaseVersionsDb");
    ns = asNamespace(pkgname);
    assign(dbNewname, txdb, envir = ns);
    namespaceExport(ns, dbNewname);

    # packageStartupMessage(AnnotationDbi:::annoStartupMessages(""))

}

# Disconnect from dbfile on onload package
# .onUnload = function(libpath)
# {
#     dbFileDisconnect(mirbaseversions_dbconn())
# }
